p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,388 @@
+	rd := protoio.NewDelimitedReader(str, maxMsgSize)
+	msg.Reset()
+	if err := rd.ReadMsg(msg); err != nil {
+		str.Reset()
+		return nil, 0, fmt.Errorf("failed to read CONNECT message from remote peer: %w", err)
+	}
+	rtt := time.Since(start)
+
+	if t := msg.GetType(); t != pb.HolePunch_CONNECT {
+		str.Reset()
+		return nil, 0, fmt.Errorf("expect CONNECT message, got %s", t)
+	}
+
+	addrs := addrsFromBytes(msg.ObsAddrs)
+
+	msg.Reset()
+	msg.Type = pb.HolePunch_SYNC.Enum()
+	if err := w.WriteMsg(msg); err != nil {
+		str.Reset()
+		return nil, 0, fmt.Errorf("failed to send SYNC message for hole punching: %w", err)
+	}
+	return addrs, rtt, nil
+}
+
+// attempts to make a direct connection with the remote peer of `relayConn` by co-ordinating a hole punch over
+// the given relay connection `relayConn`.
+func (hs *Service) HolePunch(rp peer.ID) error {
+	// short-circuit check to see if we already have a direct connection
+	for _, c := range hs.host.Network().ConnsToPeer(rp) {
+		if !isRelayAddress(c.RemoteMultiaddr()) {
+			return nil
+		}
+	}
+
+	// short-circuit hole punching if a direct dial works.
+	// attempt a direct connection ONLY if we have a public address for the remote peer
+	for _, a := range hs.host.Peerstore().Addrs(rp) {
+		if manet.IsPublicAddr(a) && !isRelayAddress(a) {
+			forceDirectConnCtx := network.WithForceDirectDial(hs.ctx, "hole-punching")
+			dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+
+			tstart := time.Now()
+			err := hs.host.Connect(dialCtx, peer.AddrInfo{ID: rp})
+			dt := time.Since(tstart)
+			cancel()
+
+			if err != nil {
+				hs.tracer.DirectDialFailed(rp, dt, err)
+				break
+			}
+			hs.tracer.DirectDialSuccessful(rp, dt)
+			log.Debugf("direct connection to peer %s successful, no need for a hole punch", rp.Pretty())
+			return nil
+		}
+	}
+
+	// hole punch
+	for i := 0; i < maxRetries; i++ {
+		addrs, rtt, err := hs.initiateHolePunch(rp)
+		if err != nil {
+			hs.handlerError(rp, err)
+			return err
+		}
+		synTime := rtt / 2
+		log.Debugf("peer RTT is %s; starting hole punch in %s", rtt, synTime)
+
+		// wait for sync to reach the other peer and then punch a hole for it in our NAT
+		// by attempting a connect to it.
+		timer := time.NewTimer(synTime)
+		select {
+		case <-timer.C:
+			pi := peer.AddrInfo{
+				ID:    rp,
+				Addrs: addrs,
+			}
+			start := time.Now()
+			hs.tracer.StartHolePunch(rp, addrs, rtt)
+			err := hs.holePunchConnect(pi, i+1)
+			dt := time.Since(start)
+			hs.tracer.EndHolePunch(rp, dt, err)
+			if err == nil {
+				log.Infof("hole punching with %s successful after %s", rp, dt)
+				return nil
+			}
+		case <-hs.ctx.Done():
+			timer.Stop()
+			return hs.ctx.Err()
+		}
+	}
+	return fmt.Errorf("all retries for hole punch with peer %s failed", rp)
+}
+
+func (hs *Service) handlerError(p peer.ID, err error) {
+	hs.tracer.ProtocolError(p, err)
+	log.Warn(err)
+}
+
+func (hs *Service) handleNewStream(s network.Stream) {
```

Comment:
We're about to dial arbitrary addresses, we should do some more checks:

1. Make sure that the connection on which we received the request is a relay.
2. Maybe make sure that it's an _outbound_ dial? It looks like we've defined the connection receiver as the side that initiates the upgrade.

Reply:
1. Makes sense, I added this check.
2. It's the peer who's not behind the relay who's initiating the upgrade, so that won't work.

Reply:
> It's the peer who's not behind the relay who's initiating the upgrade, so that won't work.

https://github.com/libp2p/go-libp2p/pull/1168/files#diff-c550e388063ff4af6c56eca2bd1e8d5f32a59ed74a481130884d8620c61a7ff8R337

That is:

1. Peer A receives an _inbound_ connection from peer B.
2. Peer A opens a new "hole punch" stream to peer B.
3. Peer B receives this stream, calling this function.
4. Peer B sees the underlying connection as an outbound connection.

(unless there are multiple connections)

Reply:
Makes sense, and it requires the attacker to be at least _a little bit_ more sophisticated than just connecting to nodes and asking them to dial others.